src/doc/contrib/src/process/working-on-cargo.md

```diff
@@ -124,6 +124,16 @@ typically finish in under 30 minutes.
 The reviewer might point out changes deemed necessary. Large or tricky changes
 may require several passes of review and changes.
 
+> **tip:** Prefer atomic commits where each commit is a single, complete, and coherent unit of work.
+> For example, if your feature work leads to renaming a module, make the rename its own commit.
+> However, adding an internal function that is unused is not complete of coherent.
+>
+> As part of your atomic commits, prefer adding tests as their own commit *before* any functionality changes.
```

Comment:
Could we have an example PR here?

Some people write a failing test first, not a test passing the CI.

Reply:
Added examples and tried to be clearer that we want the tests passing on each commit.